caddyfile: Reject directives in the place of site addresses

armadi1809 opened the pull request:
Resolves #6095  

@francislavoie from my understanding, that's the only change needed but I might be wrong. Also, let me know if you think we need more tests.

francislavoie commented:
Yeah I think we should have a test that ensures `http://handle` is still allowed. I think it will be because directives aren't registered with a scheme, but just a sanity check.

armadi1809 commented:
@francislavoie I get your point with the filename and line number but I am not entirely sure about where to move this new check since we need to loop through the block keys to determine if we are using a directive as a site address but we only have access to the `Tokens` at the segment level. Maybe I am missing something here though?

francislavoie commented:
I think we could either rework `ServerBlock` to hold `[]Token` instead of `[]string` for the site addresses (to retain the file/line along with the text), or move the errors to `func (p *parser) addresses() error {` where it actually parses the addresses initially.

armadi1809 commented:
I think I like the second option better as it feels more intuitive to perform the check there. One issue is that we don't have access to the `registeredDirectives` at the `caddyfile` package which makes me wonder if the only solution is reworking `ServerBlock`s. Unless there is another way of accessing all the caddy known directives.

francislavoie commented:
Ah right I forgot about that :cry: yeah it'll probably be better to pass down the tokens I guess. We can add some helper methods to ServerBlock to unwrap the tokens to a `[]string` if necessary.

francislavoie commented on caddyconfig/caddyfile/parse.go:
```diff
@@ -212,7 +212,7 @@ func (p *parser) addresses() error {
 
 	for {
 		tkn := p.Val()
-
+		token := p.Token()
```
I think we should rename `tkn` to "value" or something, just to differentiate it from `token`. Kinda confusing that they have different types with essentially the same name.

francislavoie commented on caddyconfig/httpcaddyfile/httptype.go:
```diff
@@ -65,11 +65,11 @@ func (st ServerType) Setup(
-				return nil, warnings, fmt.Errorf("cannot define a matcher outside of a site block: '%s'", k)
+			if j == 0 && strings.HasPrefix(k.Text, "@") {
+				return nil, warnings, fmt.Errorf("cannot define a matcher outside of a site block: '%s'", k.Text)
 			}
```
Let's add the file/line here as well

francislavoie commented:
Tests are failing - looks like we need to add the `","` trim in another spot.

armadi1809 commented:
@francislavoie I think we should be good to go now. I changed the location of that `Trim` to be where we actually add the token rather than in the helper that maps the tokens to strings. I also addressed your comments from above. Thanks for your feedback and help.

francislavoie commented on caddyconfig/caddyfile/parse.go:
```diff
@@ -211,10 +210,10 @@ func (p *parser) addresses() error {
 	for {
-		tkn := p.Val()
-
+		value := p.Val()
+		token := p.Token()
 		// special case: import directive replaces tokens during parse-time
```
Nit: the newline was there to visually separate the sections

```suggestion
		value := p.Val()
		token := p.Token()

		// special case: import directive replaces tokens during parse-time
```

francislavoie commented on caddyconfig/caddyfile/parse.go:
```diff
@@ -254,11 +253,12 @@ func (p *parser) addresses() error {
 			}
 
-			p.block.Keys = append(p.block.Keys, tkn)
+			token.Text = strings.Trim(token.Text, ",")
```
I don't think we need to trim here, since `value` already has the comma removed from the end. We can just do `token.Text = value` I think.

francislavoie commented on caddytest/integration/caddyfile_adapt/http_valid_directive_like_site_address.txt:
```diff
@@ -0,0 +1,46 @@
+http://handle {
+				file_server
```
4 tabs -> 1 tab :sweat_smile: 

```suggestion
	file_server
```

armadi1809 commented:
All the feedback items from above should now be resolved.

francislavoie reviewed (approved):
Thanks, this is great!

I made a few minor test adjustments, i.e. the error now asserts that the file/line is present, and I changed a couple other tests to fix some port conflicts that happened on my machine due to the tests using common ports (admin 2019 but I already have Caddy running, and 8080 is commonly used for some HTTP apps).
